### PR TITLE
 SILGen: Don't crash on withoutActuallyEscaping of objc blocks

### PR DESCRIPTION
**Pull request by aschwaighofer**
rdar://38689227

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test and merge

File: lib/SILGen/SILGenExpr.cpp [hunk 1]
```diff
@@ -5480,6 +5480,28 @@ RValue RValueEmitter::visitMakeTemporarilyEscapableExpr(
...
 
   auto escapingFnTy = SGF.getLoweredType(E->getOpaqueValue()->getType());
+  auto silFnTy = escapingFnTy.castTo<SILFunctionType>();
+
+  // Handle @convention(block). No withoutActuallyEscaping verification yet.
+  if (silFnTy->getExtInfo().getRepresentation() !=
+      SILFunctionTypeRepresentation::Thick) {
+    RValue rvalue;
+    auto escapingClosure =
+        SGF.emitManagedRValueWithCleanup(SGF.B.createConvertFunction(
```
Comment (gottesmm) [inline lib/SILGen/SILGenExpr.cpp]: Can you use SILGenFunction APIs instead of doing this by hand here?

File: lib/SILGen/SILGenExpr.cpp [hunk 2]
```diff
@@ -5480,6 +5480,28 @@ RValue RValueEmitter::visitMakeTemporarilyEscapableExpr(
...
+    SILGenFunction::OpaqueValueState opaqueValue{
+        escapingClosure,
+        /*consumable*/ true,
+        /*hasBeenConsumed*/ false,
+    };
+    SILGenFunction::OpaqueValueRAII pushOpaqueValue(SGF, E->getOpaqueValue(),
+                                                    opaqueValue);
+
+    // Emit the guarded expression.
+    rvalue = visit(E->getSubExpr(), C);
```
Comment (gottesmm) [inline lib/SILGen/SILGenExpr.cpp]: Why not just return this rather than assign to rvalue.

**Review by gottesmm — COMMENTED** [review]
Also, maybe refactor this function after you get in the fix? It seems like it has subroutines within it.

**Review by slavapestov — CHANGES_REQUESTED** [review]
Do we have executable tests for withoutActuallyEscaping? Might want to add one with a block also.

File: lib/SILGen/SILGenExpr.cpp (hunk 2 above)
Comment (aschwaighofer) [reply lib/SILGen/SILGenExpr.cpp]: Will cleanup in follow-up commit.

**Comment by aschwaighofer** [comment]
@slavapestov Yes there is test/Interpreter/withoutActuallyEscaping.swift

File: lib/SILGen/SILGenExpr.cpp (hunk 2 above)
Comment (gottesmm) [reply lib/SILGen/SILGenExpr.cpp]: SGTM

Comment (aschwaighofer) [reply lib/SILGen/SILGenExpr.cpp]: https://github.com/apple/swift/pull/15402

**Comment by aschwaighofer** [comment]
Executable test: https://github.com/apple/swift/pull/15406

**Pull request merged**